src/chapter_3/PDAs.md

```diff
@@ -0,0 +1,369 @@
+import { expect } from 'chai';
+
+describe('puppet', () => {
+  anchor.setProvider(anchor.Provider.env());
+
+  const puppetProgram = anchor.workspace.Puppet as Program<Puppet>;
+  const puppetMasterProgram = anchor.workspace.PuppetMaster as Program<PuppetMaster>;
+
+  const puppetKeypair = web3.Keypair.generate();
+
+  it('Does CPI!', async () => {
+    const [puppetMasterPDA, puppetMasterBump] = await web3.PublicKey
+      .findProgramAddress([], puppetMasterProgram.programId);
+
+    await puppetProgram.rpc.initialize(puppetMasterPDA, {
+      accounts: {
+        puppet: puppetKeypair.publicKey,
+        user: anchor.getProvider().wallet.publicKey,
+        systemProgram: web3.SystemProgram.programId,
+      },
+      signers: [puppetKeypair]
+    });
+
+    await puppetMasterProgram.rpc.pullStrings(puppetMasterBump, new anchor.BN(42),{
+      accounts: {
+        puppetProgram: puppetProgram.programId,
+        puppet: puppetKeypair.publicKey,
+        authority: puppetMasterPDA
+      ,
+    }});
+
+    expect((await puppetProgram.account.data
+      .fetch(puppetKeypair.publicKey)).data.toNumber()).to.equal(42);
+  });
+});
+```
+
+The `authority` is no longer a randomly generated keypair but a PDA derived from the puppet-master program. This means the puppet-master can sign with it which it does inside `pullStrings`. It's worth noting that our implementation also allows non-canonical bumps but again because we are only interesting in being able to sign we don't care which bump is used.
+
+## PDAs: Conclusion
+
+This section serves as a brief recap of all the different things you can do with PDAs.
+
+First, you can create hashmaps with them. We created a user stats PDA which was derived from the user address. This derivation linked the user address and the user stats account, allowing the latter to be easily found given the former.
+
+A subtle result of this hashmap structure is enforced uniqueness. When `init` is used with `seeds` and `bump`, it will always search for the canonical bump. This means that it can only be called once (because the 2nd time it's called the PDA will already be initialized). To illustrate how powerful enforced uniqueness is, consider a decentralized exchange program. In this program, anyone can create a new market for two assets. However, the program creators want liquidity to be concentrated so there should only be one market for every combination of two assets. This could be done without PDAs but would require a global account that saves all the different markets. Then upon market creation, the program would check whether the asset combination exists in the global market list. With PDAs this can be done much more straightforward. Any market would simply be the PDA of the mint addresses of the two assets. The program would then check whether either of the two possible PDAs (because the market could've been created with the assets in reverse order) already exists.
```

Comment:
maybe a bit much new info for a conclusion

Reply:
:shrug: